inginious/agent/docker_agent/__init__.py

```diff
@@ -567,14 +652,56 @@ async def handle_running_container(self, info: DockerRunningJob, future_results)
+                            elif msg["type"] == "run_student_command":  # We use non docker-docker communication !
+                                if student_write_stream is None:
+                                    student_container_id = msg["student_container_id"]
+                                    student_sock = await self._docker.attach_to_container(student_container_id)
+                                    student_read_stream, student_write_stream = await asyncio.open_connection(
+                                        sock=student_sock.get_socket())
+                                await self._write_to_container_stdin(student_write_stream,
+                                                                     {"type": "run_student_command",
+                                                                      "socket_id": msg["socket_id"],
+                                                                      "command": msg["command"],
+                                                                      "teardown_script": msg["teardown_script"],
+                                                                      "student_container_id": msg[
+                                                                          "student_container_id"],
+                                                                      "working_dir": msg["working_dir"],
+                                                                      "ssh": msg["ssh"],
+                                                                      "user": msg["user"]
+                                                                      })
+
+                                if not msg["ssh"]:  # classical run_student (not ssh_student) with a kata runtime -> handle student_container outputs
+                                    self._loop.create_task(self._handle_student_container_outputs(student_read_stream, write_stream))
+
+                                if msg["ssh"] and not msg["only_dockers"]:
+                                    await self.start_ssh(student_read_stream, info)  # If using ssh with kata: wait for ssh info and start ssh
+
+                            elif msg["type"] == "stdin":
+                                if student_write_stream is None:
+                                    student_container_id = msg["student_container_id"]
+                                    student_sock = await self._docker.attach_to_container(student_container_id)
+                                    student_read_stream, student_write_stream = await asyncio.open_connection(
+                                        sock=student_sock.get_socket())
+                                await self._write_to_container_stdin(student_write_stream, msg)
+
+                            elif msg["type"] == "student_signal":  # Transfer the signal to the student_container
+                                if student_write_stream is None:
```

Comment:
Same piece of code as above here. Only the forwarded message if different. Both clauses can be merged, and the forwarded message set according to ``msg["type"]``

Reply:
Indeed. Removed many lines :)